pkg/distributor/receiver/shim.go

```diff
@@ -0,0 +1,318 @@
+	streams := make(map[string]*logproto.Stream)
+	rss := ld.ResourceLogs()
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+		ill := rs.InstrumentationLibraryLogs()
+		for i := 0; i < ill.Len(); i++ {
+			logs := ill.At(i).LogRecords()
+			for k := 0; k < logs.Len(); k++ {
+				pLog := logs.At(k)
+				labels := parseLabel(rs.Resource().Attributes(), pLog.Attributes())
+				entry, err := parseEntry(pLog, format)
+				if err != nil {
+					return nil, err
+				}
+				if stream, ok := streams[labels]; ok {
+					stream.Entries = append(stream.Entries, *entry)
+					continue
+				}
+				// Add the entry as a new stream
+				streams[labels] = &logproto.Stream{
+					Labels:  labels,
+					Entries: []logproto.Entry{*entry},
+				}
+			}
+		}
+	}
+	req := &logproto.PushRequest{
+		Streams: make([]logproto.Stream, 0, len(streams)),
+	}
+
+	for _, stream := range streams {
+		req.Streams = append(req.Streams, *stream)
+	}
+
+	return req, nil
+}
+
+func parseEntry(pLog pdata.LogRecord, format string) (*logproto.Entry, error) {
+	var line string
+	if format == LogFormatLogfmt {
+		records := make(map[string]interface{})
+		records["severity_number"] = int32(pLog.SeverityNumber())
+		records["severity_text"] = pLog.SeverityText()
+		records["name"] = pLog.Name()
+		records["body"] = pLog.Body().AsString()
+		records["flags"] = pLog.Flags()
+		records["trace_id"] = pLog.TraceID().HexString()
+		records["span_id"] = pLog.SpanID().HexString()
+
+		buf := &bytes.Buffer{}
+		enc := logfmt.NewEncoder(buf)
+		keys := make([]string, 0, len(records))
+		for k := range records {
+			keys = append(keys, k)
+		}
+		sort.Strings(keys)
+		for _, k := range keys {
+			err := enc.EncodeKeyval(k, records[k])
+			if err == logfmt.ErrUnsupportedValueType {
+				err := enc.EncodeKeyval(k, fmt.Sprintf("%+v", records[k]))
+				if err != nil {
+					return nil, err
+				}
+				continue
+			}
+			if err != nil {
+				return nil, err
+			}
+		}
+		line = buf.String()
+	} else if format == LogFormatJSON {
+		record := LokiLogRecord{
+			SeverityNumber: int32(pLog.SeverityNumber()),
+			SeverityText:   pLog.SeverityText(),
+			Name:           pLog.Name(),
+			Body:           pLog.Body().AsString(),
+			Flags:          pLog.Flags(),
+			TraceID:        pLog.TraceID().HexString(),
+			SpanID:         pLog.SpanID().HexString(),
+		}
+
+		data, err := json.Marshal(&record)
+		if err != nil {
+			return nil, err
+		}
+		line = string(data)
+	} else {
+		return nil, fmt.Errorf("unsupported format type:%s", format)
+	}
+
+	return &logproto.Entry{
+		Timestamp: time.Unix(0, int64(pLog.Timestamp())),
+		Line:      line,
+	}, nil
+}
+
+func parseLabel(resource pdata.AttributeMap, attr pdata.AttributeMap) string {
+	labelMap := make([]string, 0)
+	resource.Range(func(key string, attr pdata.AttributeValue) bool {
+		labelMap = append(labelMap, fmt.Sprintf("%s=%q", key, attr.AsString()))
+		return true
+	})
+	attr.Range(func(key string, attr pdata.AttributeValue) bool {
```

Comment:
This would probably make this extra high cardinality. I understand this isn't that desirable for Loki.

Reply:
The user's label in the otel log sdk, loki needs to be saved.
 At least the code on our team needs to save the label to loki, 
 otherwise the query performance will be very poor. Especially when using json operations.

Reply:
I guess we need someone from the Loki team to decide on this :-)

cc @cyriltovena, @owen-d

Reply:
Other folks using the Loki exporter as part of the OpenTelemetry Collector prefer the other way around. Someone from the Loki dev team should really weigh in here, as they know their users better.